### DISCUSSION TITLE
FGFDMExec class seems to break when placed in a C++ array or vector

### DISCUSSION THREAD

**Discussion by scotteh432** [JSBSim-Team/jsbsim#999] [Q&A] [answered]
Hello everyone!

I'm working on a project where I want to run N number of aircraft flight dynamic models simultaneously, and write JSBSim property values about their position to each other on every iteration, so that the autopilot and systems can make react based on the positions of the other aircraft. However I'm running into issues with the FGFDMExec class in the C++ API. For some reason, once I place the class inside an array, the code breaks when I execute the RunIC() or Run() class methods. 

Here's a code sample of the issue distilled down to the relevant parts. 

```
#include <FGFDMExec.h>
#include <vector>

int main(int argc, char* argv[])
{

// Some code here to retrieve the inputs to the relevant portions for this question 
std::vector<string> aircraftFolders = {// vector of multiple aircraftFolder paths, one for each model}
std::vector<string> systemFolders = {// vector of multiple systemFolder paths, one for each model}
std::vector<string> engineFolders = {// vector of multiple engineFolder paths, one for each model}
std::vector<string>outputFolders = {// vector of multiple outputFolder paths, one for each model}
std::vector<string>scriptFiles = {// vector of multiple scriptPath paths, one for each model}

// Relevant code for my question
  std::vector<JSBSim::FGFDMExec> vectorOfFdms;
  for (size_t i = 0; i < aircraftFolders.size(); ++i) {
    JSBSim::FGFDMExec x;
    x.SetDebugLevel(0);
    x.SetAircraftPath(SGPath(aircraftFolders[i]));
    x.SetSystemsPath(SGPath(systemFolders[i]));
    x.SetEnginePath(SGPath(engineFolders[i]));
    x.SetOutputPath(SGPath(outputFolders[i]));
    x.LoadScript(SGPath(scriptFiles[i]));
    x.Setdt(0.001);
    x.RunIC();
    vectorOfFdms.push_back(x);
  }

  bool result = true;
  while (result) {
    for (size_t i = 0; i < vectorOfFdms.size(); ++i) {
      result = vectorOfFdms[i].Run(); // <-- exception is thrown here for some reason
    }
  }
}
```

For some reason, an exception is thrown when attempting to call the instances Run() method once they are inside the vector. This also happens if I run RunIC() once the instance is inside the vector. However it has no problem accessing any other method such as GetAircraftPath(). This problem also occurs if I use arrays instead of vectors. 

I can avoid this problem by not using arrays / vectors. I can create a separate variable for each FGFDMExec class, and then run multiple cases simultaneously just fine (although this does not let me create N arbitrary aircraft models).  So the issue isn't with defining multiple FGFDMExec classes.  For some reason, it seems that the FGFDMExec class is breaking once placed in array. My C++ knowledge is quite limited so there may be something obvious staring me in the face. 

Any help would be much appreciated.

**Comment by seanmcleod** [comment]
Take a look at - https://stackoverflow.com/questions/2275076/is-stdvector-copying-the-objects-with-a-push-back

**Reply by scotteh432** [reply, marked as answer]
Fixed it. After reading the link above I understand a little bit better about the distinctions between pointing reference, and copying. Pre-initializing the vector solved the issue.

```
  std::vector < JSBSim::FGFDMExec> vectorOfFdms(aircraftFolders.size());
  for (size_t i = 0; i < aircraftFolders.size(); ++i) {
    vectorOfFdms[i].SetDebugLevel(0);
    vectorOfFdms[i].SetAircraftPath(SGPath(aircraftFolders[i]));
    vectorOfFdms[i].SetSystemsPath(SGPath(systemFolders[i]));
    vectorOfFdms[i].SetEnginePath(SGPath(engineFolders[i]));
    vectorOfFdms[i].SetOutputPath(SGPath(outputFolders[i]));
    vectorOfFdms[i].LoadScript(SGPath(scriptFiles[i]));
    vectorOfFdms[i].Setdt(timeStepSize);
  }
```
I can't wait to get back into my Python comfort zone, and away from all this scary C++ stuff.
Thanks for your help :)
